os/fs/smartfs/smartfs_utils.c

```diff
@@ -1065,209 +1163,276 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
-			}
-		}
-	} else {
-		/* Use the provided sector number */
+	/*Read the entire logical sector containing the old entry into fs_rwbuffer*/
+	readwrite.logsector = psector;
+	readwrite.count = fs->fs_llformat.availbytes;
+	readwrite.offset = 0;
+	readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
 
-		nextsector = sectorno;
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Error %d reading sector %d data\n", ret, new_entry.dsector);
+		goto errout;
 	}
 
+	entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
+
 	/* Create the directory entry to be written in the parent's sector */
 
 #ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
-	smartfs_wrle16(&entry->firstsector, nextsector);
+	smartfs_wrle16(&entry->firstsector, new_entry.firstsector);
 	smartfs_wrle16(&entry->utc, time(NULL));
 #else
-	entry->firstsector = nextsector;
+	entry->firstsector = new_entry.firstsector;
 	entry->utc = time(NULL);
 #endif
 	memset(entry->name, 0, fs->fs_llformat.namesize);
 	strncpy(entry->name, filename, fs->fs_llformat.namesize);
+	entry->flags = new_entry.flags;
 
 	/* Now write the new entry to the parent directory sector */
-
 	readwrite.logsector = psector;
 	readwrite.offset = offset;
 	readwrite.count = entrysize;
 	readwrite.buffer = (uint8_t *)&fs->fs_rwbuffer[offset];
+	fdbg("\nWriting new entry to sector = %d, offset = %d, count = %d\n", psector, offset, entrysize);
+	fdbg("\nWriting flags = %x\n", entry->flags);
 	ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long) &readwrite);
 	if (ret < 0) {
-		fdbg("failed to write new entry to parent directory psector : %d\n", psector);
+		fdbg("failed to write new entry to parent directory psector : %d\n", new_entry.dsector);
 		goto errout;
 	}
 
 	/* Now fill in the entry */
-
-	direntry->firstsector = nextsector;
-	direntry->dsector = psector;
-	direntry->doffset = offset;
+	ret_entry->firstsector = new_entry.firstsector;
+	ret_entry->dsector = psector;
+	ret_entry->doffset = offset;
 #ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
-	direntry->flags = smartfs_rdle16(&entry->flags);
-	direntry->utc = smartfs_rdle32(&entry->utc);
+	ret_entry->flags = smartfs_rdle16(&entry->flags);
+	ret_entry->utc = smartfs_rdle32(&entry->utc);
 #else
-	direntry->flags = entry->flags;
-	direntry->utc = entry->utc;
+	ret_entry->flags = entry->flags;
+	ret_entry->utc = entry->utc;
 #endif
-	direntry->datlen = 0;
-	if (direntry->name == NULL) {
-		direntry->name = (FAR char *)kmm_malloc(fs->fs_llformat.namesize + 1);
-		if (direntry->name == NULL) {
+	ret_entry->datlen = 0;
+	if (ret_entry->name == NULL) {
+		ret_entry->name = (FAR char *)kmm_malloc(fs->fs_llformat.namesize + 1);
+		if (ret_entry->name == NULL) {
 			ret = ERROR;
 			goto errout;
 		}
 	}
 
-	memset(direntry->name, 0, fs->fs_llformat.namesize + 1);
-	strncpy(direntry->name, filename, fs->fs_llformat.namesize);
+	memset(ret_entry->name, 0, fs->fs_llformat.namesize + 1);
+	strncpy(ret_entry->name, filename, fs->fs_llformat.namesize);
```

Comment:
yeah.. this also need to be checked.
I think this should be done when it created.

Reply:
I didn't get this comment sorry?